logic/tool/TypeInference.java

```diff
@@ -194,7 +194,9 @@ private FunctionalIterator<Map<Identifier.Variable.Name, Label>> typePermutation
             Map<Retrievable, Order> sortOrder = new HashMap<>();
             sortVars.forEach(var -> sortOrder.put(var, Order.Asc.ASC));
             traversal.modifiers().filter(inferenceFilter).sorting(Modifiers.Sorting.create(sortVars, sortOrder));
-            return traversalEng.iterator(traversal).map(vertexMap -> {
+            return traversalEng.iterator(traversal).filter(vertexMap ->
+                !iterate(vertexMap.map().entrySet()).anyMatch(e -> thingInferenceVars().contains(e.getKey()) && e.getValue().asType().isAbstract())
```

Comment:
note: we filter out permutations containing abstract 'thing' variable types in type inference permutations, which are used when validating rules' insertability: we can guarantee that thing variables will never provide instances of abstract types, so we can eliminate these permutations.